### PR TITLE
Ice cream revamp: Ice cream is a component and cones can hold multiple servings of ice cream now.

### LINKED ISSUE #45868: Ice cream cart can dispense invalid flavor waffle cones
Title, href abusing an ice cream cart allows for custom cone flavors that aren't supported, plus some other oddities.

## Reproduction:
- Stand next to the chef's ice cream cart found in the kitchen backroom/freezer
- Open the menu and change the cone dispense number href to something that isn't vanilla or chocolate
(Should be 3 to 6, there might also be some side issue where setting it to 0 grabs some unlisted stock of vanilla, best not to ask why)

- Abuse byond with your rotten hrefs
- Get forbidden cone flavors that are too much for the universe to visualize

Secretspice

Additionally from just memory there are some other effects like sending the amount of vanilla in the machine to negative infinity while trying to create an infinite amount of waffle cones or something. Probably some other stuff I forgot too, so have fun if you plan to mess with it yourself.

> Like most places our ice cream machine doesn't work, unlike those places ours is so badly broken it doesn't refrain from shattering physics because you pressed the wrong buttons and pushed the lever instead of pulling it.

### PR DESCRIPTION
**Pull request by Ghommie**
## About The Pull Request
Title. I always wanted to renovate that dusty ice cream code a little. 
Ice cream flavo(u)rs are now datums and you can stack up to three scoops of ice cream on a cone (or more, with VV). I worked through the night on this, but I have also tested it, so I can say it works swimmingly.
EDIT: Ice cream is now a component called ice_cream_holder.

## Why It's Good For The Game
Doing justice to those people who want more servings of ice cream in their cones. Also refurbishing the code. This will fix #45868.

TODO in a next future:
- Ice cream parlor venue datum for robot patrons. Maybe.

## Changelog
:cl:
add: Sundae, honkdae and space freezy can be filled with ice cream from the ice cream vat.
add: Ice creams cones, as well as the aforementioned treats, can hold multiple servings of ice cream now, up to three.
refactor: All because ice cream has been refactored into a component, and its flavours made into datums.
/:cl:

### CODE REVIEW

**Comment by ArcaneDefence** [comment]
I'm terrified and excited

**Comment by Farquaar** [comment]
Are the extra scoops visible on the sprite?

**Comment by Ghommie** [comment]
> 
> 
> Are the extra scoops visible on the sprite?

Yessir, they are overlays.

**Comment by coiax** [comment]
I love this, but I'll need some time to review it all.

File: code/__DEFINES/dcs/signals.dm [hunk 1]
```diff
@@ -1003,6 +1003,10 @@
 ///from base of datum/component/two_handed/proc/unwield(mob/living/carbon/user): (/mob/user)
 #define COMSIG_TWOHANDED_UNWIELD "twohanded_unwield"
 
+// /datum/component/ice_cream_holder signals.
+/// Adds a serving of ice cream to the object. Bool. Refer to __DEFINES/food.dm for flavour defines: (flavour, reagents, custom_name, forced)
+#define COMSIG_ICE_CREAM_ADD_SERVING "ice_cream_add_serving"
```
Comment (AnturK) [inline code/__DEFINES/dcs/signals.dm:1008]: This seems completely redundant, it's only used to initialize the borg/crematorium types. And the signal is so specific it's not very useful.

File: code/datums/components/food/ice_cream_holder.dm [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+
+/datum/component/ice_cream_holder/proc/on_examine_more(atom/source, mob/mob, list/examine_list)
+	var/scoops_len = length(scoops)
+	if(scoops_len == 1 || length(uniqueList(scoops)) == 1) /// Only one flavour.
+		var/key = scoops[1]
+		var/datum/ice_cream_flavour/flavour = GLOB.ice_cream_flavours[LAZYACCESS(special_scoops, key) || key]
+		if(flavour?.desc) //I scream.
+			examine_list += "[source.p_theyre(TRUE)] filled with scoops of [flavour ? flavour.name : "broken, unhappy"] icecream."
+		else
+			examine_list += replacetext(replacetext("[source.p_theyre(TRUE)] [flavour.desc]", "$CONE_NAME", initial(source.name)), "$CUSTOM_NAME", key)
```
Comment (AnturK) [inline code/datums/components/food/ice_cream_holder.dm:112]: Unrelated to the pr but we could use some generic madlibs helpers.

File: code/datums/components/food/ice_cream_holder.dm [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+///Adds a new flavour to the ice cream cone.
+/datum/ice_cream_flavour/proc/add_flavour(datum/component/ice_cream_holder/target, datum/reagents/R, custom_name)
+	var/atom/owner = target.parent
+	LAZYADD(target.scoops, custom_name || name)
+	if(icon_state)
+		LAZYADD(target.scoop_overlays, icon_state)
+	if(custom_name)
+		LAZYSET(target.special_scoops, custom_name, name)
+
+	owner.reagents.maximum_volume += EXTRA_MAX_VOLUME_PER_SCOOP
```
Comment (AnturK) [inline code/datums/components/food/ice_cream_holder.dm:205]: This looks abusable if the edible is extractable ?

File: code/modules/food_and_drinks/kitchen_machinery/icecream_vat.dm [hunk 4]
```diff
@@ -149,50 +105,59 @@
...
 	if(..())
 		return
 	if(href_list["select"])
-		dispense_flavour = text2num(href_list["select"])
-		flavour_name = get_flavour_name(dispense_flavour)
-		src.visible_message("<span class='notice'>[usr] sets [src] to dispense [flavour_name] flavoured ice cream.</span>")
+		var/datum/ice_cream_flavour/flavour = GLOB.ice_cream_flavours[href_list["select"]]
+		if(!flavour || flavour.hidden) //Nice try, tex.
+			return
+		src.visible_message("<span class='notice'>[usr] sets [src] to dispense [href_list["select"]] flavoured ice cream.</span>")
```
Comment (AnturK) [inline code/modules/food_and_drinks/kitchen_machinery/icecream_vat.dm:133]: no need for src.

File: code/datums/components/food/ice_cream_holder.dm (hunk 3 above)
Comment (Ghommie) [reply code/datums/components/food/ice_cream_holder.dm:205]: Slightly, but most edibles are not extractable via syringe, plus that would count more as a creative usage of a food container than else (it's probably not a big deal, but i don't want to jinx it).

File: code/datums/components/food/ice_cream_holder.dm (hunk 2 above)
Comment (Ghommie) [reply code/datums/components/food/ice_cream_holder.dm:112]: I suppose, though I'd like an example later to get a better idea of this suggestion.

File: code/__DEFINES/dcs/signals.dm (hunk 1 above)
Comment (Ghommie) [reply code/__DEFINES/dcs/signals.dm:1008]: So do you suggest I use `GetComponent()` and handle it without signals instead?

Comment (AnturK) [reply code/__DEFINES/dcs/signals.dm:1008]: For this case i'd suggest creating presets for the robo/crematorium cones that start with the given flavours.

File: code/datums/components/food/ice_cream_holder.dm (hunk 2 above)
Comment (AnturK) [reply code/datums/components/food/ice_cream_holder.dm:112]: Oh don't worry, this is just idle thought. I was thinking some simple `adlib(text,replacement_dictionary/datum)`

**Pull request merged**